### PR TITLE
Group zeroes when printing binary terms

### PR DESCRIPTION
**Pull request by essen**
This is a proposal to change the way binaries are formatted. There are of course more optimization opportunities like adding more clauses to the `zeroes` function. I did not try to run tests yet, I will look into it once I know this has a chance of being merged.

The main reason I want this is to reduce the amount of info printed when I trace the Cowboy tests (they use a lot of very large binaries full of zeroes). But this would also help me reduce the amount of info printed when tracing HTTP/2 code in the normal case.

Instead of printing potentially very large number of zeroes
separated by commas:

``` erlang
1> io:format("~p~n", [<<1, 0:8000, 1>>]).
<<1,0,0,0,0,0,0,0,0,...
... many lines of 0s...
  0,0,0,0,1>>
```

It is more efficient and often more readable to group them
together:

``` erlang
1> io:format("~p~n", [<<1, 0:8000, 1>>]).
<<1,0:8000,1>>
```

The gain in efficiency of course mostly comes from writing
much smaller printable binaries to the terminal or files,
but we also benefit from a reduction in memory usage.

### CODE REVIEW

File: lib/stdlib/src/io_lib.erl [hunk 1]
```diff
@@ -449,6 +449,16 @@ write_binary_body(B, D, T, Acc) when D =:= 1; T =:= 0->
     {["..."|Acc], B};
 write_binary_body(<<X:8>>, _D, _T, Acc) ->
     {[integer_to_list(X)|Acc], <<>>};
+write_binary_body(<<0:24,Rest0/bitstring>>, D, T, Acc) ->
```
Comment (hauleth) [inline lib/stdlib/src/io_lib.erl]: It would probably be worth mentioning that this will fire only if there are 3 or more zeroed bytes.

Comment (essen) [reply lib/stdlib/src/io_lib.erl]: Right. It's because `0:16,` is longer than `0,0,`. But it's an arbitrary decision nevertheless.

**Comment by vans163** [comment]
Great feature to have but not default please. I oppose this because often when doing packet analysis your copy+pasting the binaries around and you wanna that process to be smooth (into hex-editor etc).

**Comment by essen** [comment]
I would say it should be the default. Packet analysis is not what most people are doing when reading output that include binaries, many times the binary is there simply because it's in some state or trace or other and we don't need to look at the binaries specifically.

I don't disagree that sometimes it's nice to be able to compare, but in that case perhaps the hex formatter that `ssl` is using should be made more readily available and you can just call that with your binary as argument. When I was doing a lot of packet analysis many years ago I was just logging everything into files and then opened those files in an Hex editor and wrote programs to analyze the packets, I did not need to look at the formatted binaries from inside Erlang. I do when debugging sometimes but in those cases grouping zeroes help more than hinder.

**Comment by vans163** [comment]
> I would say it should be the default. Packet analysis is not what most people are doing when reading output that include binaries, many times the binary is there simply because it's in some state or trace or other and we don't need to look at the binaries specifically.
> 
> I don't disagree that sometimes it's nice to be able to compare, but in that case perhaps the hex formatter that `ssl` is using should be made more readily available and you can just call that with your binary as argument. When I was doing a lot of packet analysis many years ago I was just logging everything into files and then opened those files in an Hex editor and wrote programs to analyze the packets, I did not need to look at the formatted binaries from inside Erlang. I do when debugging sometimes but in those cases grouping zeroes help more than hinder.

Its much faster to analyse in realtime using the REPL without writing to files. One of the beauties of the VM is the REPL is just so damn powerful.  If this somehow does become the default though, it would confuse a lot of people, but regardless a way to disable it then would be nice to not have to keep a patchset around.

**Comment by hauleth** [comment]
@vans163 I do not get it. How it will be harder to analyse output in form of:

```
<<1, 2, 0:160, 3>>
```

Than

```
<<1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3>>
```

Because I do not get it. For me it is much easier to analyse former.

**Comment by vans163** [comment]
> @vans163 I do not get it. How it will be harder to analyse output in form of:
> 
> ```
> <<1, 2, 0:160, 3>>
> ```
> 
> Than
> 
> ```
> <<1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3>>
> ```
> 
> Because I do not get it. For me it is much easier to analyse former.
```
<<0:24, 1, 0, 0, 1, 2>>
<<0, 0, 1, 0:24, 2>>
<<0:24, 1, 0, 1, 0:40, 5, 7, 1>>
```
VS
```
<<0, 0, 0, 1, 0, 0, 1, 2>>
<<0, 0, 1, 0, 0, 0, 2>>
<<0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 5, 7, 1>>
```
Copy [0:24, 1, 0, 1, 0:40, 5, 7, 1], Paste in hex_editor/idapython/etc [error invalid decimal]

Is this not obvious?

**Comment by ferd** [comment]
I can never debug binaries using the shell because:

1. it would be very dangerous not to turn on truncating. Any large data dump can destabilize a whole node, particularly if you try to print out stuff like a 10mb image in RAM to the binary notation
2. `<<1,0,0,0,0,0,0,0,0,...>>` is necessarily what I get when truncating
3. `<<0, 0, 1, 0:24, 2>>` lets me see more of the data I need when truncating without risking overload of the system

I'm +1 on this PR.

**Comment by vans163** [comment]
> I can never debug binaries using the shell because:
> 
> 1. it would be very dangerous not to turn on truncating. Any large data dump can destabilize a whole node, particularly if you try to print out stuff like a 10mb image in RAM to the binary notation
> 2. `<<1,0,0,0,0,0,0,0,0,...>>` is necessarily what I get when truncating
> 3. `<<0, 0, 1, 0:24, 2>>` lets me see more of the data I need when truncating without risking overload of the system
> 
> I'm +1 on this PR.

This is not the usecase! The usecase is when you specifically print out the binaries as your looking to analyze them. Not just random analysis on dumps of bins that get printed due to crashes or other prints.  As long as there is a way to print it like normal, and sure exceptions/etc VM wide can have the new format. 

Lets see what OTP team says.

**Comment by ferd** [comment]
I'd just expect  `~w` to print them as they do right now, which would objectively make more sense for analysis anyway rather than being stuck with arbitrary column-formatting from the VM, where some of the binary data might turn into strings anyway. That's your default mode if you really need it?

**Comment by llelf** [comment]
> @vans163 I do not get it. How it will be harder to analyse output in form of:
> <<1, 2, 0:160, 3>>
> Than
> <<1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3>>
> Because I do not get it. For me it is much easier to analyse former.

It’s significantly easier to see the size and the placement of (non-0) data in the latter case.
But I’m ±0 on the PR.

**Comment by garazdawi** [comment]
One other thing which is I think is a "good thing" about this change is that if a large binary needs to be logged and thus trimmed in order to not blow up the logs, this change has the possibility to make the log contain much more relevant data. that is

    2019-11-19T13:46:19.651421+01:00 error: <<0:800, 1>>

instead of

    2019-11-19T13:46:19.651421+01:00 error: <<0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...>>

(and yes, we haven't forgotten about this PR)

**Comment by josevalim** [comment]
I agree with @vans163 that `<<0:24, 1, 0, 1, 0:40, 5, 7, 1>>` is harder to read than `<<0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 5, 7, 1>>`. I can easily count when I have four zeroes or more in a row, so reading `0, 0, 0, 0` is easier than `0:32`, especially when mixed with others numbers without the size modifier.

But at some point that is no longer the case. So maybe a reasonable compromise is to increase the minimum number of zeros that trigger collapsing? I would pick 16 (for no other reason than it is the first power of 2 above 10) and call it a day.

**Comment by essen** [comment]
It should just be configurable, perhaps via a kernel/stdlib environment parameter. Personally I've been dealing with the binary syntax for so long that `0:32` is a LOT easier to read than four zeroes. And 16 would make the patch do nothing in far too many scenarios. 16 zeroes isn't that common while 4 is very common.

**Comment by eproxus** [comment]
Not sure if it would be technically possible, but I would love for the truncation to be in the style of `<<1,0,0,0,0,0,0,0,0,...:298>>`.

**Comment by essen** [comment]
The change for this would need to be done here so it's definitely technically possible: https://github.com/erlang/otp/pull/2378/files#diff-b9ef2de127d85fb8f57ab477c95cbc2eL449

Don't know about whether it's worth it though. No opinion.

**Comment by hauleth** [comment]
@eproxus and how it should be handled in case when after `0` there is a non-zero?

```
<<1,0,0,0,0,0,0,0,0,...:298,1>>
```

What should be the limit of `0` to be displayed before going with `:<num>`?

**Comment by garazdawi** [comment]
We have decided to not include this change. We do not see enough benefit to counter the downsides and the extra complexity this brings.

Some of the factors that influenced this decision was:
* There would still not be any extra information available. The depth limitation would work exactly as it does today as it is not only a limitation in how much is printed but also how much is processed.
* It is relatively rare to have many zeroes in a row in protocols and when there are many zeroes (like a float or 64-bit integer) the zeroes would sometimes make things more confusing. e.g. The original binary `<<256:64,1:64>>` could be printed as `<<0:48,1,0:64,1>>` or `<<0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1>>`.
* It would depend too much upon the protocol if it is useful or not, very much like the ascii printouts of `hexdump -C`.
* Since this is not an obvious improvement for everyone we have opted to take the conservative route.

We have also decided to not add any configuration or modifier for this option as that would introduce even more complexity into something that already is very complex.

Thank you for the PR!

**Comment by essen** [comment]
Oh well. At least it's easy enough to patch in when necessary. Thanks for the consideration!

**Comment by hauleth** [comment]
@garazdawi how about squashing only full `<<0:64>>`, so only if more than 8 or 16 zeros in the row would be squashed. In the same way it is done in `hexdump -C`:

```
$ ruby -e 'print "\x00"*32' | hexdump -C
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000020
```

So your example would still be readable, but it would allow squashing really large sets of identical bytes.

**Comment by garazdawi** [comment]
> it would allow squashing really large sets of identical bytes.

It would allow squashing of large sets of zeroes, which I would argue is not very common.

**Comment by essen** [comment]
> * There would still not be any extra information available. The depth limitation would work exactly as it does today as it is not only a limitation in how much is printed but also how much is processed.

For the record, note that in my case, this PR is mostly useful in cases where no depth limitation is applied and I either do not have control over the code that prints the stuff; or when I can control it it involves some rather cumbersome solution, for example setting up a custom `dbg` tracer that does a limited depth print.

The extra information could be nice to have but it's not really my concern.

**Comment by garazdawi** [comment]
> For the record, note that in my case, this PR is mostly useful in cases where no depth limitation is applied

Yes, that was clear to us. Others that have commented in this PR (including me....) thought that this would affect the term depth printing, which I wanted to point out that it would not.

> setting up a custom dbg tracer that does a limited depth print

When I'm done with #2475 you should be able to do this much more simply :)

**Comment by eproxus** [comment]
@hauleth Well, the `...:Length` could be any data, just as `...` is any data in squashed Erlang terms today. I didn't mean for `...` to only replace zeroes, but any binary data.

**Pull request closed without merging**